Separate exit code for error

**Tell us about your environment**

* **ESLint Version:** Any
* **Node Version:** 6
* **npm Version:** None, I use Yarn latest.

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

Not relevant.

**What did you do? Please include the actual source code causing the issue.**

Running an eslint instance which should crash (config error, intern failing like a permission issue...)

**What did you expect to happen?**

Having a different exist code issue.

**What actually happened? Please include the actual, raw output from ESLint.**

The exit code issue is the same as a well running eslint with rule violation detection.


Currently, eslint give the same exit code when some lint rules are not respected and when the binary is failing (example: When the provided configuration file does not exists).

On a automated script, we can't see the difference.

It would be great to have different exit codes:

* `1` For not respected lint rules.
* `2` When eslint is failing

Thanks.

Another example with [`yamllint`](https://github.com/adrienverge/yamllint):

```
$ yamllint app/config/config.yml 
app/config/config.yml
  1:1       warning  missing document start "---"  (document-start)
  2:8       error    too many spaces inside braces  (braces)
  2:33      error    too many spaces inside braces  (braces)
  3:8       error    too many spaces inside braces  (braces)
  3:31      error    too many spaces inside braces  (braces)
  4:8       error    too many spaces inside braces  (braces)
  4:29      error    too many spaces inside braces  (braces)
  5:8       error    too many spaces inside braces  (braces)
  5:37      error    too many spaces inside braces  (braces)
  6:8       error    too many spaces inside braces  (braces)
  6:42      error    too many spaces inside braces  (braces)
  7:8       error    too many spaces inside braces  (braces)
  7:31      error    too many spaces inside braces  (braces)
  8:8       error    too many spaces inside braces  (braces)
  8:28      error    too many spaces inside braces  (braces)
  8:31      warning  too few spaces before comment  (comments)
  9:8       error    too many spaces inside braces  (braces)
  9:28      error    too many spaces inside braces  (braces)
  9:31      warning  too few spaces before comment  (comments)
  10:8      error    too many spaces inside braces  (braces)
  10:26     error    too many spaces inside braces  (braces)
  19:6      warning  missing starting space in comment  (comments)
$ echo $?
1
$ yamllint 
usage: yamllint [-h] [-c CONFIG_FILE | -d CONFIG_DATA]
                [-f {parsable,standard}] [-s] [-v]
                FILE_OR_DIR [FILE_OR_DIR ...]
yamllint: error: too few arguments
$ echo $?
2
```

--- a/bin/eslint.js
+++ b/bin/eslint.js
@@ -53,7 +53,7 @@ process.once("uncaughtException", err => {
         console.error(err.stack);
     }
 
-    process.exitCode = 1;
+    process.exitCode = 2;
 });
 
 if (useStdIn) {

--- a/docs/user-guide/command-line-interface.md
+++ b/docs/user-guide/command-line-interface.md
@@ -450,3 +450,11 @@ ESLint supports `.eslintignore` files to exclude files from the linting process 
     **/vendor/*.js
 
 A more detailed breakdown of supported patterns and directories ESLint ignores by default can be found in [Configuring ESLint](configuring.md#ignoring-files-and-directories).
+
+## Exit codes
+
+ESLint will exit with one of the following exit codes:
+
+* `0`: Linting was successful and there are no linting errors. If the `--max-warnings` flag is set to `n`, the number of linting warnings is at most `n`.
+* `1`: Linting was successful and there is at least one linting error, or there are more linting warnings than allowed by the `--max-warnings` option.
+* `2`: Linting was unsuccessful due to a configuration problem or an internal error.

--- a/docs/user-guide/command-line-interface.md
+++ b/docs/user-guide/command-line-interface.md
@@ -453,7 +453,7 @@ A more detailed breakdown of supported patterns and directories ESLint ignores b
 
 ## Exit codes
 
-ESLint will exit with one of the following exit codes:
+When linting files, ESLint will exit with one of the following exit codes:
 
 * `0`: Linting was successful and there are no linting errors. If the `--max-warnings` flag is set to `n`, the number of linting warnings is at most `n`.
 * `1`: Linting was successful and there is at least one linting error, or there are more linting warnings than allowed by the `--max-warnings` option.

--- a/lib/cli.js
+++ b/lib/cli.js
@@ -139,7 +139,7 @@ const cli = {
             currentOptions = options.parse(args);
         } catch (error) {
             log.error(error.message);
-            return 1;
+            return 2;
         }
 
         const files = currentOptions._;
@@ -153,11 +153,11 @@ const cli = {
         } else if (currentOptions.printConfig) {
             if (files.length) {
                 log.error("The --print-config option must be used with exactly one file name.");
-                return 1;
+                return 2;
             }
             if (useStdin) {
                 log.error("The --print-config option is not available for piped-in code.");
-                return 1;
+                return 2;
             }
 
             const engine = new CLIEngine(translateOptions(currentOptions));
@@ -176,12 +176,12 @@ const cli = {
 
             if (currentOptions.fix && currentOptions.fixDryRun) {
                 log.error("The --fix option and the --fix-dry-run option cannot be used together.");
-                return 1;
+                return 2;
             }
 
             if (useStdin && currentOptions.fix) {
                 log.error("The --fix option is not available for piped-in code; use --fix-dry-run instead.");
-                return 1;
+                return 2;
             }
 
             const engine = new CLIEngine(translateOptions(currentOptions));
@@ -207,7 +207,7 @@ const cli = {
 
                 return (report.errorCount || tooManyWarnings) ? 1 : 0;
             }
-            return 1;
+            return 2;
 
 
         }

--- a/tests/bin/eslint.js
+++ b/tests/bin/eslint.js
@@ -138,7 +138,7 @@ describe("bin/eslint.js", () => {
                     ["--stdin"], { cwd: "/", env: { HOME: "/" } }
                 );
 
-                const exitCodePromise = assertExitCode(child, 1);
+                const exitCodePromise = assertExitCode(child, 2);
                 const stderrPromise = getOutput(child).then(output => {
                     assert.match(
                         output.stderr,
@@ -316,7 +316,7 @@ describe("bin/eslint.js", () => {
     describe("handling crashes", () => {
         it("prints the error message to stderr in the event of a crash", () => {
             const child = runESLint(["--rule=no-restricted-syntax:[error, 'Invalid Selector [[[']", "Makefile.js"]);
-            const exitCodeAssertion = assertExitCode(child, 1);
+            const exitCodeAssertion = assertExitCode(child, 2);
             const outputAssertion = getOutput(child).then(output => {
                 const expectedSubstring = "Syntax error in selector";
 
@@ -330,7 +330,7 @@ describe("bin/eslint.js", () => {
         it("prints the error message pointing to line of code", () => {
             const invalidConfig = `${__dirname}/../fixtures/bin/.eslintrc.yml`;
             const child = runESLint(["--no-ignore", invalidConfig]);
-            const exitCodeAssertion = assertExitCode(child, 1);
+            const exitCodeAssertion = assertExitCode(child, 2);
             const outputAssertion = getOutput(child).then(output => {
                 const expectedSubstring = "Error: bad indentation of a mapping entry at line";
 

--- a/tests/lib/cli.js
+++ b/tests/lib/cli.js
@@ -121,7 +121,7 @@ describe("cli", () => {
             const filePath = getFixturePath("files");
             const result = cli.execute(`--blah --another ${filePath}`);
 
-            assert.strictEqual(result, 1);
+            assert.strictEqual(result, 2);
         });
 
     });
@@ -245,7 +245,7 @@ describe("cli", () => {
             const filePath = getFixturePath("passing.js");
             const exit = cli.execute(`-f fakeformatter ${filePath}`);
 
-            assert.strictEqual(exit, 1);
+            assert.strictEqual(exit, 2);
         });
     });
 
@@ -265,14 +265,14 @@ describe("cli", () => {
             const filePath = getFixturePath("passing.js");
             const exit = cli.execute(`-f ${formatterPath} ${filePath}`);
 
-            assert.strictEqual(exit, 1);
+            assert.strictEqual(exit, 2);
         });
     });
 
     describe("when executing a file with a lint error", () => {
         it("should exit with error", () => {
             const filePath = getFixturePath("undef.js");
-            const code = `--no-ignore --config --rule no-undef:2 ${filePath}`;
+            const code = `--no-ignore --rule no-undef:2 ${filePath}`;
 
             const exit = cli.execute(code);
 
@@ -308,7 +308,7 @@ describe("cli", () => {
 
     describe("when executing with version flag", () => {
         it("should print out current version", () => {
-            cli.execute("-v");
+            assert.strictEqual(cli.execute("-v"), 0);
 
             assert.strictEqual(log.info.callCount, 1);
         });
@@ -316,7 +316,7 @@ describe("cli", () => {
 
     describe("when executing with help flag", () => {
         it("should print out help", () => {
-            cli.execute("-h");
+            assert.strictEqual(cli.execute("-h"), 0);
 
             assert.strictEqual(log.info.callCount, 1);
         });
@@ -403,7 +403,7 @@ describe("cli", () => {
             assert.throws(() => {
                 const exit = cli.execute(code);
 
-                assert.strictEqual(exit, 1);
+                assert.strictEqual(exit, 2);
             }, /Error while loading rule 'custom-rule': Cannot read property/);
         });
 
@@ -559,7 +559,7 @@ describe("cli", () => {
 
             const exit = cli.execute(code);
 
-            assert.strictEqual(exit, 1);
+            assert.strictEqual(exit, 2);
             assert.isTrue(log.info.notCalled);
             assert.isTrue(log.error.calledOnce);
         });
@@ -572,7 +572,7 @@ describe("cli", () => {
 
             const exit = cli.execute(code);
 
-            assert.strictEqual(exit, 1);
+            assert.strictEqual(exit, 2);
             assert.isTrue(log.info.notCalled);
             assert.isTrue(log.error.calledOnce);
         });
@@ -611,7 +611,7 @@ describe("cli", () => {
             const filePath = getFixturePath("passing.js");
             const exit = cli.execute(`--no-ignore --parser-options test111 ${filePath}`);
 
-            assert.strictEqual(exit, 1);
+            assert.strictEqual(exit, 2);
         });
 
         it("should exit with no error if parser is valid", () => {
@@ -862,7 +862,7 @@ describe("cli", () => {
 
             const exitCode = localCLI.execute("--fix .", "foo = bar;");
 
-            assert.strictEqual(exitCode, 1);
+            assert.strictEqual(exitCode, 2);
         });
 
     });
@@ -1020,7 +1020,7 @@ describe("cli", () => {
 
             const exitCode = localCLI.execute("--fix --fix-dry-run .", "foo = bar;");
 
-            assert.strictEqual(exitCode, 1);
+            assert.strictEqual(exitCode, 2);
         });
     });
 
@@ -1042,7 +1042,7 @@ describe("cli", () => {
 
             assert.isTrue(log.info.notCalled);
             assert.isTrue(log.error.calledOnce);
-            assert.strictEqual(exitCode, 1);
+            assert.strictEqual(exitCode, 2);
         });
 
         it("should error out when executing on text", () => {
@@ -1050,7 +1050,7 @@ describe("cli", () => {
 
             assert.isTrue(log.info.notCalled);
             assert.isTrue(log.error.calledOnce);
-            assert.strictEqual(exitCode, 1);
+            assert.strictEqual(exitCode, 2);
         });
     });
 

